# Credit Card Fraud Detection

---

In [1]:
# Dependencies
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [5]:
# load data in a Pandas dataframe
credit_card_df = pd.read_csv('./dataset/creditcard.csv')

In [6]:
# Number of rows and columns in the dataset
credit_card_df.shape

(284807, 31)

In [20]:
# Display first 5 rows of the dataset
credit_card_df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [21]:
# Display last 5 rows of the dataset
credit_card_df.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [22]:
# dataset information
credit_card_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [23]:
# Number of missing values in each column
credit_card_df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [24]:
# Checking the distribution of the data
credit_card_df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

### The data is pretty much unbalanced
#### 0 - Normal Transaction
#### 1 - Fraudulent Transaction

In [9]:
credit_card_arr = credit_card_df.values

In [10]:
credit_card_arr.shape

(284807, 31)

In [14]:
credit_card_arr.mean(axis=0)

array([ 9.48138596e+04,  1.16597995e-15,  3.41690805e-16, -1.37314996e-15,
        2.08686861e-15,  9.60406632e-16,  1.49010721e-15, -5.55646730e-16,
        1.17755593e-16, -2.40645529e-15,  2.23975129e-15,  1.67332693e-15,
       -1.25499520e-15,  8.17603027e-16,  1.20629628e-15,  4.91300284e-15,
        1.43766606e-15, -3.80011269e-16,  9.57213260e-16,  1.03981681e-15,
        6.40670259e-16,  1.65656173e-16, -3.44485005e-16,  2.57864790e-16,
        4.47196822e-15,  5.34091469e-16,  1.68709835e-15, -3.66645261e-16,
       -1.22040449e-16,  8.83496193e+01,  1.72748563e-03])

In [15]:
credit_card_arr.std(axis=0)

array([4.74880626e+04, 1.95869237e+00, 1.65130568e+00, 1.51625234e+00,
       1.41586609e+00, 1.38024431e+00, 1.33226875e+00, 1.23709143e+00,
       1.19435081e+00, 1.09863016e+00, 1.08884785e+00, 1.02071124e+00,
       9.99199635e-01, 9.95272483e-01, 9.58593928e-01, 9.15314405e-01,
       8.76251349e-01, 8.49335573e-01, 8.38174738e-01, 8.14039072e-01,
       7.70923671e-01, 7.34522725e-01, 7.25700286e-01, 6.24459199e-01,
       6.05646005e-01, 5.21277155e-01, 4.82226167e-01, 4.03631786e-01,
       3.30082685e-01, 2.50119670e+02, 4.15271167e-02])

### We can see that the mean of all the 28 features is 0 and std is 1 approximately

In [25]:
# Separating the legit and fraudulent transactions for analysis (Statistical analysis on the data points)
legit = credit_card_df[credit_card_df['Class'] == 0]
fraud = credit_card_df[credit_card_df['Class'] == 1]

In [26]:
print(legit.shape, fraud.shape)

(284315, 31) (492, 31)


In [28]:
legit['Amount'].describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [29]:
fraud['Amount'].describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

### We can see the mean of fraudulent transactions is 122.21 whereas mean of legit transactions is 88.29

In [30]:
# Compae mean values for both classes
credit_card_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

## Because of the highly unbalanced nature of the dataset, we perform
# UNDER-SAMPLING

#### Build a sample dataset containing similar distribution of normal and fraudulent transactions

In [31]:
legit_sample = legit.sample(n=492)

In [32]:
print(legit.shape, legit_sample.shape, fraud.shape)

(284315, 31) (492, 31) (492, 31)


In [33]:
# Concatenate the two dataframes
new_df = pd.concat([legit_sample, fraud], axis=0)

In [34]:
new_df.shape

(984, 31)

In [35]:
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
197014  131803.0 -0.367835  0.481206 -1.730261 -0.397471  1.899645 -1.614545   
192043  129512.0 -1.537584  1.280091 -0.080053 -0.954728  1.554181  5.587238   
260487  159579.0 -0.608001  1.345322 -0.661349 -0.711060  0.671579 -0.672292   
85360    60742.0  1.091314  0.048305  0.325065  1.029046 -0.114520  0.127445   
154010  100846.0  1.903731 -0.687862 -0.091386  0.283846 -0.655592  0.556059   

              V7        V8        V9  ...       V21       V22       V23  \
197014  1.369031 -0.476501 -0.441386  ...  0.102510  0.526829 -0.040734   
192043 -3.914806 -5.276837  0.520300  ...  2.547703 -1.048995 -0.238385   
260487  0.603493  0.421504 -0.423440  ... -0.146861 -0.497753  0.187868   
85360  -0.127941  0.189414 -0.070474  ... -0.006634 -0.138222 -0.041027   
154010 -1.221686  0.312379  2.644343  ... -0.089461 -0.059237  0.305879   

             V24       V25       V26       V27       V28  Amount  Class  
197014  0.665880 -0.736565  0.381131  0.331907  0.181018   88.46      0  
192043  0.660310  1.211897  0.028698  0.299298  0.239369   13.35      0  
260487  0.654255 -1.067418 -0.019419  0.157605  0.233531    4.49      0  
85360  -0.347827  0.392435 -0.416407  0.021152  0.014200   41.00      0  
154010  0.216920 -0.657703  0.420069 -0.060298 -0.049761   39.00      0  

[5 rows x 31 columns]

In [36]:
new_df['Class'].value_counts()

1    492
0    492
Name: Class, dtype: int64

In [37]:
new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92526.097561  0.186008 -0.081027  0.034799  0.024215 -0.000195   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.051361 -0.007537 -0.021192  0.076766  ... -0.012747 -0.014426   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.059586  0.015372 -0.009313 -0.002639 -0.034862 -0.006900 -0.017907   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       89.668537  
1      122.211321  

[2 rows x 30 columns]

#### Split the data into features and target

In [38]:
X = new_df.drop(columns=['Class'], axis=1)
Y = new_df['Class']

In [39]:
print(X.shape, Y.shape)

(984, 30) (984,)


In [40]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [41]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [42]:
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(787, 30)
(197, 30)
(787,)
(197,)


### Model Training
#### Logistic Regression

In [43]:
model = LogisticRegression()

In [44]:
model.fit(X_train, Y_train)

LogisticRegression()

### Model Evaluation
#### Training data Accuracy score

In [45]:
X_train_predictions = model.predict(X_train)

In [46]:
X_train_predictions.shape

(787,)

In [47]:
training_data_accuracy = accuracy_score(X_train_predictions, Y_train)

In [50]:
print('Training data accuracy:', training_data_accuracy)

Training data accuracy: 0.9466327827191868


#### Testing data Accuracy score

In [52]:
X_test_predictions = model.predict(X_test)

In [53]:
print(X_test_predictions.shape)

(197,)


In [54]:
testing_data_accuracy = accuracy_score(X_test_predictions, Y_test)

In [55]:
print('Testing data accuracy:', testing_data_accuracy)

Testing data accuracy: 0.9187817258883249
